smart-contract-extensions/contracts/KeyPurchaser.sol

```diff
@@ -0,0 +1,162 @@
+
+/**
+ * @notice Purchase a key priced in any ERC-20 token - either once or as a regular subscription.
+ * This allows the user to purchase or subscribe to a key with 1 tx (`approve`)
+ * or if the token supports it, with 1 signed message (`permit`).
+ *
+ * The user can remove approval to cancel anytime.
+ *
+ * Risk: if the user transfers or cancels the key, they would naturally expect that also cancels
+ * the subscription but it does not. This should be handled by the frontend.
+ */
+contract KeyPurchaser is Initializable, Stoppable
+{
+  using Address for address payable;
+  using SafeERC20 for IERC20;
+
+  // set on initialize and cannot change
+  IPublicLock public lock;
+  uint public maxKeyPrice;
+  uint public maxPurchaseCount;
+  uint public renewWindow;
+  uint public renewMinFrequency;
+  bool public isSubscription;
+
+  // admin can change these anytime
+  string public name;
+  bool internal disabled;
+
+  // store minimal history
+  mapping(address => uint) public timeOfLastPurchase;
+
+  /**
+   * @notice Called once to set terms that cannot change later on.
+   */
+  function initialize(
+    IPublicLock _lock,
+    uint _maxKeyPrice,
+    uint _renewWindow,
+    uint _renewMinFrequency,
+    bool _isSubscription
+  ) public
+    initializer()
+  {
+    require(_renewWindow > 0 || !_isSubscription, 'INVALID_RENEW_WINDOW');
+    require(_renewMinFrequency > 0 || !_isSubscription, 'INVALID_RENEW_MINFREQUENCY');
+
+    _initializeAdminRole();
+    lock = _lock;
+    maxKeyPrice = _maxKeyPrice;
+    renewWindow = _renewWindow;
+    renewMinFrequency = _renewMinFrequency;
+    isSubscription = _isSubscription;
+    approveSpending();
+  }
+
+  /**
+   * @notice Approves the lock to spend funds held by this contract.
+   * @dev Automatically called on initialize, needs to be called again if the tokenAddress changes.
+   * No permissions required, it's okay to call this again.
+   */
+  function approveSpending() public
+  {
+    IERC20 token = IERC20(lock.tokenAddress());
+    if(address(token) != address(0))
+    {
+      token.approve(address(lock), uint(-1));
+    }
+  }
+
+  /**
+   * @notice Used by admins to update metadata which may be leveraged by the frontend.
+   */
+  function config(
+    string memory _name,
+    bool _disabled
+  ) public
+    onlyAdmin()
+  {
+    name = _name;
+    disabled = _disabled;
+  }
+
+  /**
+   * @notice Indicates if this purchaser should be exposed as an option to users on the frontend.
+   * False does not necessarily mean previous subs will no longer work (see `stopped` for that).
+   */
+  function isActive() public view returns(bool)
+  {
+    return !stopped() && !disabled;
+  }
+
+  /**
+   * @notice Called by anyone to purchase or renew a key on behalf of a user.
+   * The user must have ERC-20 spending approved and the purchase must meet the terms
+   * defined during initialization.
+   */
+  function purchaseFor(
+    address payable _user,
+    address _referrer,
```

Comment:
I wonder if _recipient(or _beneficiary) would be more clear than _user here?

Reply:
k renamed